OpenOversight/app/commands.py

```diff
@@ -411,6 +421,388 @@ def bulk_add_officers(filename, no_create, update_by_name, update_static_fields)
+                if job_id is None:
+                    raise Exception(
+                        "Job title {} not found for department.".format(
+                            row["job_title"]
+                        )
+                    )
+                row["job_id"] = job_id
+                officer = all_officers.get(row["officer_id"])
+                if not officer:
+                    raise Exception(
+                        "Officer with id {} does not exist (in this department)".format(
+                            row["officer_id"]
+                        )
+                    )
+                row["officer_id"] = officer.id
+                _create_or_update_model(
+                    row=row,
+                    existing_model_lookup=id_to_assignment,
+                    create_method=create_assignment_from_dict,
+                    update_method=update_assignment_from_dict,
+                    force_create=force_create,
+                    model=Assignment,
+                )
+                counter += 1
+                if counter % 1000 == 0:
+                    print("Processed {} assignments.".format(counter))
+        print("Done with assignments. Processed {} rows.".format(counter))
+
+    counter = 0
+    if salaries_csv is not None:
+        with open(salaries_csv) as f:
+            csv_reader = csv.DictReader(f)
+            csv_reader.fieldnames = _unify_field_names(csv_reader.fieldnames)
+            _check_provided_fields(
+                csv_reader,
+                required_fields=["id", "officer_id", "salary", "year"],
+                optional_fields=["overtime_pay", "is_fiscal_year"],
+                csv_name="salaries",
+            )
+            existing_salaries = (
+                Salary.query.join(Salary.officer)
+                .filter(Officer.department_id == department_id)
+                .all()
+            )
+            id_to_salary = {salary.id: salary for salary in existing_salaries}
+            for row in csv_reader:
+                officer = all_officers.get(row["officer_id"])
+                if not officer:
+                    raise Exception(
+                        "Officer with id {} does not exist (in this department)".format(
+                            row["officer_id"]
+                        )
+                    )
+                row["officer_id"] = officer.id
+                _create_or_update_model(
+                    row=row,
+                    existing_model_lookup=id_to_salary,
+                    create_method=create_salary_from_dict,
+                    update_method=update_salary_from_dict,
+                    force_create=force_create,
+                    model=Salary,
+                )
+                counter += 1
+                if counter % 1000 == 0:
+                    print("Processed {} salaries.".format(counter))
+        print("Done with salaries. Processed {} rows.".format(counter))
```

Comment:
Some of the code looks relatively familiar between salary/officer/assignment. Can these update functions be further generified?

Reply:
Yeah, I wasn't too excited with how this turned out looking, but I also couldn't come up with a better way to extract more common functionality without making it look weird, but I will give it another try